File: Python/Dynamic Programming/L_frogJump.py
from typing import List


def canCross(stones: List[int]) -> bool:
        d = dict((x,set()) for x in stones)
        if stones[1] != 1:
            return False
        d[1].add(1)
        for i in range(len(stones[1:])):
            for j in d[stones[i]]:
                for k in range(j-1, j+2):
                    if k > 0 and stones[i]+k in d:
                        d[stones[i]+k].add(k)
        return d[stones[-1]] != set()

File: cpp/Dynamic Programming/L_frogJump.cpp
#include <bits/stdc++.h>
using namespace std;

bool canCross(vector<int> &stones)
{
    int n = stones.size();

    std::unordered_map<int, std::unordered_set<int>> dp;
    for (int stone : stones)
    {
        dp[stone] = std::unordered_set<int>();
    }
    dp[0].insert(0);

    for (int i = 0; i < n; i++)
    {
        for (int k : dp[stones[i]])
        {
            for (int step = k - 1; step <= k + 1; step++)
            {
                if (step != 0 && dp.count(stones[i] + step))
                {
                    dp[stones[i] + step].insert(step);
                }
            }
        }
    }

    return !dp[stones.back()].empty();
}